component/metrics/receiver.go

```diff
@@ -13,9 +14,51 @@ type Receiver struct {
 // RiverCapsule marks receivers as a capsule.
 func (r Receiver) RiverCapsule() {}
 
-// FlowMetric is a wrapper around a single metric without the timestamp
+// FlowMetric is a wrapper around a single metric without the timestamp.
 type FlowMetric struct {
-	GlobalRefID uint64
-	Labels      labels.Labels
-	Value       float64
+	globalRefID uint64
+	labels      labels.Labels
+	value       float64
+}
+
+// NewFlowMetric instantiates a new flow metric
+func NewFlowMetric(globalRefID uint64, lbls labels.Labels, value float64) *FlowMetric {
+	// Always ensure we have a valid global ref id
+	if globalRefID == 0 {
+		globalRefID = GlobalRefMapping.GetOrAddGlobalRefID(lbls)
+	}
+	return &FlowMetric{
+		globalRefID: globalRefID,
+		labels:      lbls,
+		value:       value,
+	}
+}
+
+// GlobalRefID Retrieves the GlobalRefID
+func (fw *FlowMetric) GlobalRefID() uint64 { return fw.globalRefID }
+
+// Value returns the value
+func (fw *FlowMetric) Value() float64 { return fw.value }
+
+// LabelsCopy returns a copy of the labels structure
+func (fw *FlowMetric) LabelsCopy() labels.Labels {
+	return fw.labels.Copy()
+}
+
+// RawLabels returns the actual underlying labels that SHOULD be treated as immutable. Usage of this
+// must be very careful to ensure that nothing that consume this mutates labels in anyway.
+func (fw *FlowMetric) RawLabels() labels.Labels {
+	return fw.labels
+}
+
+// Relabel applies normal prometheus relabel rules and returns a flow metric. NOTE this may return itself.
+func (fw *FlowMetric) Relabel(cfgs ...*promrelabel.Config) *FlowMetric {
+	retLbls := promrelabel.Process(fw.labels)
+	if retLbls == nil {
+		return nil
+	}
+	if retLbls.Hash() == fw.labels.Hash() {
+		return fw
+	}
```

Comment:
There's a really minuscule chance of a hash collision here, I'd also make sure that if the hashes match the labels are also identical.

Reply:
Also entirely forgot to use cfgs...

Reply:
Fixed and added tests.